### DISCUSSION TITLE
How to guarantee that a `phase` belongs to a the current generation?

### DISCUSSION THREAD

**Discussion by WillSewell** [argoproj/argo-rollouts#2873] [Q&A]
We want to track the progress of a rollout. We do this by first polling the `Rollout.status.workloadGeneration` field until it matches the generation returned when the pod spec was updated.

Then we poll the `Rollout.status.phase` until it is either Degraded/Paused/Healthy (i.e. no longer Progressing).

The problem is, the update of `status.workloadGeneration` and the update of `status.phase` isn't atomic in all cases. So it's possible that we read the new `status.workloadGeneration` but the `status.phase` from the previous generation.

Example controller logs where it first updates `status.workloadGeneration` and then updates `status.phase`:

```
time="2023-07-07T09:27:10Z" level=info msg="Patched: {\"status\":{\"abort\":null,\"abortedAt\":null,\"currentPodHash\":\"674487bbb9\",\"observedGeneration\":\"3199\",\"workloadObservedGeneration\":\"1812\"}}" generation=3199 namespace=example re
sourceVersion=13927768850 rollout=example
time="2023-07-07T09:27:10Z" level=info msg="Patched conditions: {\"status\":{\"conditions\":[{\"lastTransitionTime\":\"2023-07-05T15:22:21Z\",\"lastUpdateTime\":\"2023-07-05T15:22:21Z\",\"message\":\"Rollout is paused\",\"reason\":\"RolloutPause
d\",\"status\":\"False\",\"type\":\"Paused\"},{\"lastTransitionTime\":\"2023-07-07T09:04:24Z\",\"lastUpdateTime\":\"2023-07-07T09:04:24Z\",\"message\":\"RolloutCompleted\",\"reason\":\"RolloutCompleted\",\"status\":\"False\",\"type\":\"Completed
\"},{\"lastTransitionTime\":\"2023-07-07T09:09:57Z\",\"lastUpdateTime\":\"2023-07-07T09:09:57Z\",\"message\":\"Rollout has minimum availability\",\"reason\":\"AvailableReason\",\"status\":\"True\",\"type\":\"Available\"},{\"lastTransitionTime\":
\"2023-07-07T09:27:10Z\",\"lastUpdateTime\":\"2023-07-07T09:27:10Z\",\"message\":\"Retrying Rollout after abort\",\"reason\":\"RolloutRetry\",\"status\":\"Unknown\",\"type\":\"Progressing\"}],\"message\":\"more replicas need to be updated\",\"ph
ase\":\"Progressing\"}}" generation=3199 namespace=example resourceVersion=13927768892 rollout=example
```

Given this, I'm not sure how we can guarantee the phase we are reading is part of the desired generation. Is there some way of doing this, or an alternative way of satisfying our use case that we might not have considered? We could workaround this by adding some delay between reading the desired generation, and polling for the `status.phase, but I'd rather not if there was a less brittle alternative 🙂

**Comment by tompreston** [comment]
Hey Will appreciate this is dead old, so not expecting a response, but what do you mean about this comment?

> The problem is, the update of status.workloadGeneration and the update of status.phase isn't atomic in all cases

In the latest version of Argo v1.7.2 it looks like the update [is an atomic patch operation](https://github.com/argoproj/argo-rollouts/blob/ce345e6a431b6f44776ef7453559b1d00dec9b85/rollout/sync.go#L776). It looks the same in an older version too v1.2.2.

**Reply by WillSewell** [reply]
Hey Tom, just from skimming the code, I think the issue is that while the update is atomic in `persistRolloutStatus`, it is _not_ atomic in `patchCondition` because this only updates `Phase` but not `WorkloadObservedGeneration`.

But given it's been a while since I reported this, I'd recommend gathering some empirical evidence that this is actually an issue in the latest version of Argo Rollouts.
